### ISSUE TITLE
Terraform apply rebuilds all components when a map of resources passed to a module has changed

### ISSUE THREAD

**Issue by JoeButler99** [hashicorp/terraform#26320]
### Terraform Version

```
Terraform v0.13.3
+ provider registry.terraform.io/hashicorp/azuread v1.0.0
+ provider registry.terraform.io/hashicorp/azurerm v2.28.0
+ provider registry.terraform.io/hashicorp/random v2.3.0
```

### Terraform Configuration Files

```terraform

# An initial module (module.aad_users_and_groups) Creates the Azure AD users and groups.
# It is passed a map of group names to create, and returns the resource object map as an output

resource "azuread_group" "aad_groups" {
  for_each = var.aad_groups
  name     = each.value.name
}

output "aad_groups" {
  value = azuread_group.aad_groups
}

# In the same apply, this ouptut is fed to a module which creates role assignments on a given Azure subscription, and looks up the group object id from the output aad_groups

provider "azurerm" {
  features {}
  alias           = "test_env"
  subscription_id = "d74d8784-9e9b-4518-80d1-efb4ee78970f"
}

module "test_env_roles" {
  source           = "app.terraform.io/ORGNAME/role-assignment/azuread"
  version          = "0.0.2"
  role_assignments = var.role_assignments["test_env"].assignments
  aad_groups       = module.aad_users_and_groups.aad_groups
  providers = {
    azurerm = azurerm.test_env
  }
}

# Finally the role assignment module runs over the map

resource "azurerm_role_assignment" "assignments" {
  for_each             = var.role_assignments
  principal_id         = lookup(var.aad_groups, each.value.group_key).object_id
  scope                = each.value.scope
  role_definition_name = each.value.role_definition_name
}

```

### Expected Behavior
First I should point out that this does actually work, Terraform will converge on the required state, however it seems unable to determine which role assignments if any are affected by the change in the aad_groups output so will delete and recreate them all. 
I would expect that Terraform is able to determine which elements need to change and not re-create all resources. However I suspect as it has detected the output has changed it determines it must recreate all resources which reference it, instead of only those elements of the map which are affected. 

### Actual Behavior

The plan (and subsequent apply) will re-create all resources which reference the changed aad_groups.
The change in outputs however does show that only one map element is to be added.
In the below plan, I add just a new group "sharepoint_admins2" to the var.aad_groups map. It will then if applied create this group, but also recreate all the azurerm_role_assignment resources as they use this output to determine the group object id

```pre
  # module.aad_users_and_groups.azuread_group.aad_groups["sharepoint_admins2"] will be created
  + resource "azuread_group" "aad_groups" {
      + id                      = (known after apply)
      + members                 = (known after apply)
      + name                    = "sharepoint_admins2"
      + object_id               = (known after apply)
      + owners                  = (known after apply)
      + prevent_duplicate_names = false
    }

  # module.test_env_roles.azurerm_role_assignment.assignments["test_one"] must be replaced
-/+ resource "azurerm_role_assignment" "assignments" {
      ~ id                               = "/subscriptions/00000000-9e9b-4518-80d1-0000000000/providers/Microsoft.Authorization/roleAssignments/9e3bfdcc-7e29-5db1-021b-861eb7f42523" -> (known after apply)
      ~ name                             = "9e3bfdcc-7e29-5db1-021b-861eb7f42523" -> (known after apply)
      ~ principal_id                     = "cbd50f7a-289f-44b9-bccd-3aff2b659009" -> (known after apply) # forces replacement
      ~ principal_type                   = "Group" -> (known after apply)
      ~ role_definition_id               = "/subscriptions/00000000-9e9b-4518-80d1-0000000000/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c" -> (known after apply)
        role_definition_name             = "Contributor"
        scope                            = "/subscriptions/00000000-9e9b-4518-80d1-0000000000"
      + skip_service_principal_aad_check = (known after apply)
    }

  # module.test_env_roles.azurerm_role_assignment.assignments["test_one2"] must be replaced
-/+ resource "azurerm_role_assignment" "assignments" {
      ~ id                               = "/subscriptions/00000000-9e9b-4518-80d1-0000000000/providers/Microsoft.Authorization/roleAssignments/58385ee5-d595-6a16-06fa-a2b9997eac23" -> (known after apply)
      ~ name                             = "58385ee5-d595-6a16-06fa-a2b9997eac23" -> (known after apply)
      ~ principal_id                     = "cbd50f7a-289f-44b9-bccd-3aff2b659009" -> (known after apply) # forces replacement
      ~ principal_type                   = "Group" -> (known after apply)
      ~ role_definition_id               = "/subscriptions/00000000-9e9b-4518-80d1-0000000000/providers/Microsoft.Authorization/roleDefinitions/acdd72a7-3385-48ef-bd42-f606fba81ae7" -> (known after apply)
        role_definition_name             = "Reader"
        scope                            = "/subscriptions/00000000-9e9b-4518-80d1-0000000000"
      + skip_service_principal_aad_check = (known after apply)
    }

  # module.test_env_roles.azurerm_role_assignment.assignments["users_group"] must be replaced
-/+ resource "azurerm_role_assignment" "assignments" {
      ~ id                               = "/subscriptions/00000000-9e9b-4518-80d1-0000000000/providers/Microsoft.Authorization/roleAssignments/b779bfd3-94c9-a630-984b-f55913366c47" -> (known after apply)
      ~ name                             = "b779bfd3-94c9-a630-984b-f55913366c47" -> (known after apply)
      ~ principal_id                     = "1aaa46c2-b950-4342-8f59-47093fcc1245" -> (known after apply) # forces replacement
      ~ principal_type                   = "Group" -> (known after apply)
      ~ role_definition_id               = "/subscriptions/00000000-9e9b-4518-80d1-0000000000/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c" -> (known after apply)
        role_definition_name             = "Contributor"
        scope                            = "/subscriptions/00000000-9e9b-4518-80d1-0000000000"
      + skip_service_principal_aad_check = (known after apply)
    }

Plan: 4 to add, 0 to change, 3 to destroy.

Changes to Outputs:
  ~ aad_groups = {
        sharepoint_admins  = {
            description             = null
            id                      = "5eeb7c9d-0352-45b9-b8dc-89b3a71c1031"
            members                 = []
            name                    = "sharepoint_admins"
            object_id               = "5eeb7c9d-0352-45b9-b8dc-89b3a71c1031"
            owners                  = []
            prevent_duplicate_names = false
        }
      + sharepoint_admins2 = {
          + description             = null
          + id                      = (known after apply)
          + members                 = (known after apply)
          + name                    = "sharepoint_admins2"
          + object_id               = (known after apply)
          + owners                  = (known after apply)
          + prevent_duplicate_names = false
        }
        sharepoint_users   = {
            description             = null
            id                      = "bb35d36d-166e-4522-826d-c1311d63e05b"
            members                 = []
            name                    = "sharepoint_users"
            object_id               = "bb35d36d-166e-4522-826d-c1311d63e05b"
            owners                  = []
            prevent_duplicate_names = false
        }
        testGroup          = {
            description             = null
            id                      = "cbd50f7a-289f-44b9-bccd-3aff2b659009"
            members                 = []
            name                    = "test"
            object_id               = "cbd50f7a-289f-44b9-bccd-3aff2b659009"
            owners                  = []
            prevent_duplicate_names = false
        }
      ~ users_group        = {
            description             = null
            id                      = "1aaa46c2-b950-4342-8f59-47093fcc1245"
          ~ members                 = [
              + "cfa21a42-2e0a-49ac-b867-a505028fb2da",
            ]
            name                    = "users_group"
            object_id               = "1aaa46c2-b950-4342-8f59-47093fcc1245"
            owners                  = []
            prevent_duplicate_names = false
        }
    }

```

### Additional Context
Its already a little complicated as we need to use the azure_ad provider to manage the azure tenant, and then an azure_rm provider for each subscription we manage. I understand this is down to the Azure design, not Terraform itself. In this situation, the only way I can currently think of having a fully dynamic system, would be to write templates to generate all the Terraform resources individually. This is instead of using map(object) type variables passed around inside Terraform. This approach would also allow me to generate the provider blocks dynamically, but I'd rather avoid this complex step if possible.

**Comment by JoeButler99** [comment]
Quick update - I've made it work how I expect in the module, by removing the lookup and replacing it with a direct map access.
So I have a working system. There may be a point to consider if lookup should function the same as the direct access, but if this is expected, then this is neither a bug or a feature request. 

so
-  principal_id         = lookup(var.aad_groups, each.value.group_key).object_id
becomes: 
+  principal_id         = var.aad_groups[each.value.group_key].object_id

**Comment by apparentlymart** [comment]
Hi @JoeButler99! Thanks for reporting this.

I believe this happens because the `lookup` function is implemented to return "unknown" (what the plan output calls "`(known after apply)`") if there are any unknown values anywhere in the given map:

https://github.com/hashicorp/terraform/blob/f2f84003ee8ec7ee23445322be31ed4e0e30d91c/lang/funcs/collection.go#L318-L320

This feels overly conservative to me: it really only need do that if the top-level map _itself_ is unknown (`.IsKnown` instead of `.IsWhollyKnown`), which would then make it behave like the built-in index operation. I suspect this conservative behavior persists from this function's implementation in Terraform 0.11 and earlier, where a partially-unknown collection was not representable.

Using the `lookup` function with only two arguments (that is, without a default fallback value) is a legacy form that we retain only for compatibility with older Terraform versions, because the index syntax is now preferred, but this quirk of unknown value handling affects the three-argument form too, and that is still in common use as a way to do a conditional map lookup.

---

Here's a smaller reproduction case that only needs the `hashicorp/null` provider:

```hcl
resource "null_resource" "in" {
}

locals {
  map_with_unknown = tomap({
    foo = null_resource.in.id
    bar = "bar"
  })
}

resource "null_resource" "out_lookup" {
  triggers = {
    lookup = lookup(local.map_with_unknown, "bar")
  }
}

resource "null_resource" "out_index" {
  triggers = {
    index  = local.map_with_unknown["bar"]
  }
}
```

With the dev build of Terraform 0.14 I'm using here, I get the following output:

```
Terraform will perform the following actions:

  # null_resource.in will be created
  + resource "null_resource" "in" {
      + id = (known after apply)
    }

  # null_resource.out_index will be created
  + resource "null_resource" "out_index" {
      + id       = (known after apply)
      + triggers = {
          + "index" = "bar"
        }
    }

  # null_resource.out_lookup will be created
  + resource "null_resource" "out_lookup" {
      + id       = (known after apply)
      + triggers = (known after apply)
    }
```

The correct behavior would be for `triggers` to be a known value in `null_resource.out_lookup`, matching the `triggers` for `null_resource.out_index`.

---

The above is also exhibiting a quirk in the legacy SDK which current versions of the `null` provider are built with, where a `triggers` map containing an unknown value reduces to a wholly-unknown map. That's an SDK bug rather than a Terraform Core bug, but it's a well-known one and so I suggest ignoring that oddity for the sake of this issue: the SDK team will address it in a later version.

```
2020/09/24 15:50:12 [WARN] Provider "registry.terraform.io/hashicorp/null" produced an invalid plan for null_resource.out_lookup, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .triggers: planned value cty.UnknownVal(cty.Map(cty.String)) does not match config value cty.MapVal(map[string]cty.Value{"lookup":cty.UnknownVal(cty.String)})
```

**Comment by OwenTuz** [comment]
I've had a go at implementing this in the PR above, please let me know if it makes any sense :)

**Comment by pselle** [comment]
This is merged and the fix will be in Terraform v0.14: https://github.com/hashicorp/terraform/pull/26427
